File: guests/python/textarea.py
"""The textarea conformance scene, Python port. See
guests/rust/textarea.rs and tools/scenes/textarea.steps."""

import sys

import kaya

app = kaya.App()


def count(text):
    return "0 lines" if not text else f"{len(text.splitlines())} lines"


def on_edit(text):
    lines.set(count(text))


def on_clear():
    editor.clear()
    editor.focus()


with app.window(title="textarea"):
    lines = kaya.signal("0 lines")
    with kaya.column():
        editor = kaya.textarea(on_change=on_edit)
        kaya.label(bind=lines)
        kaya.button("clear", on_click=on_clear)

sys.exit(app.run())

File: guests/js/textarea.ts
// The textarea conformance scene (tools/scenes/textarea.steps).

import * as kaya from "kaya-gui";

const app = new kaya.App();

function count(text: string): string {
  return text === "" ? "0 lines" : `${text.split("\n").length} lines`;
}

function onEdit(text: string): void {
  lines.set(count(text));
}

function onClear(): void {
  editor.clear();
  editor.focus();
}

const { lines, editor } = app.window({ title: "textarea" }, () => {
  const lines = kaya.signal("0 lines");
  let editor!: kaya.Widget;
  kaya.column(() => {
    editor = kaya.textarea({ onChange: onEdit });
    kaya.label({ bind: lines });
    kaya.button("clear", { onClick: onClear });
  });
  return { lines, editor };
});

app.run();
